api/pkg/handler/node_v1.go

```
@@ -588,3 +588,165 @@ func decodeGetNodeForCluster(c context.Context, r *http.Request) (interface{}, e
+		if err != nil {
+			return nil, kubernetesErrorToHTTPError(err)
+		}
+
+		// TODO:
+		// normally we have project, user and sshkey providers
+		// but here we decided to use machineClient and kubeClient directly to access the user cluster.
+		//
+		// how about moving machineClient and kubeClient to their own provider ?
+		machineClient, err := clusterProvider.GetMachineClientForCustomerCluster(cluster)
+		if err != nil {
+			return nil, kubernetesErrorToHTTPError(err)
+		}
+
+		dc, found := dcs[cluster.Spec.Cloud.DatacenterName]
+		if !found {
+			return nil, fmt.Errorf("unknown cluster datacenter %s", cluster.Spec.Cloud.DatacenterName)
+		}
+
+		nodeset := &req.Body
+
+		if nodeset.Spec.Template.Cloud.Openstack == nil &&
+			nodeset.Spec.Template.Cloud.Digitalocean == nil &&
+			nodeset.Spec.Template.Cloud.AWS == nil &&
+			nodeset.Spec.Template.Cloud.Hetzner == nil &&
+			nodeset.Spec.Template.Cloud.VSphere == nil &&
+			nodeset.Spec.Template.Cloud.Azure == nil {
+			return nil, errors.NewBadRequest("cannot create node without cloud provider")
+		}
+
+		//TODO: We need to make the kubelet version configurable but restrict it to master version
+		if nodeset.Spec.Template.Versions.Kubelet != "" {
+			kversion, err := semver.NewVersion(nodeset.Spec.Template.Versions.Kubelet)
+			if err != nil {
+				return nil, fmt.Errorf("failed to parse kubelet version: %v", err)
+			}
+			c, err := semver.NewConstraint(kubeletVersionConstraint)
+			if err != nil {
+				return nil, fmt.Errorf("failed to parse kubelet constraint version: %v", err)
+			}
+
+			if !c.Check(kversion) {
+				return nil, fmt.Errorf("kubelet version does not fit constraint. Allowed %s", kubeletVersionConstraint)
+			}
+			nodeset.Spec.Template.Versions.Kubelet = kversion.String()
+		} else {
+			//TODO: rework the versions
+			nodeset.Spec.Template.Versions.Kubelet = cluster.Spec.Version
+		}
+
+		if nodeset.Name == "" {
+			nodeset.Name = "kubermatic-" + cluster.Name + "-" + rand.String(5)
```

Shouldn't we check if that name is already taken?

---

Yes, I will do it.

---

What for?
If its already taken, the apiserver will tell us, and we can pass through that error to the user.
There's no benefit it adding one additional API call before, just to make sure.

We can check the error with `IsAlreadyExists`.
So we can return the same error message as if we checked before.

---

@mrIncompetent So you want endusers to see an error `Can not create NodeGroup: MachineDeployment machinedeployment-$rand-$rand already exists` and afterwards have them do the exact same thing again, this time without an error?

---

Ah right, you're referring to the random name generation! Missed that.

But yes, i would avoid the extra complexity for that. Using 5 random chars here leads to ~14 million possibilities. But not a strong opinion

---

14 milion is a lot. Not sure about it anymore. Same thing is done for nodes at the moment.

---

I'd just like to avoid the chance of random failures. AFAIK it should also be possible to use `metadata.GenerateName` which basically has this exact purpose: Generate a random name that is not used yet.

Node names are currently the same as machine names, so as long as we have no duplicate machines nodes will be unique.

---

Great idea! `metadata.GenerateName` would be perfect!

---

Yeah, seems like a perfect fit. I have adjusted generating names of machine deployments but also for machines. Thanks again for an idea!

Sample name that is generated: `machinedeployment-kubermatic-x766wbbq4q-8xsfg-f965b978d-wxz78`, where `x766wbbq4q` part is the cluster ID.